### ISSUE TITLE
Enhancement suggestion: navigation tree could reflect the currently displayed item

### ISSUE THREAD

**Issue by aaronforshaw** [MauroDataMapper/mdm-ui#219]
User can click on the navigation tree to display an item such as a model or class. The selected item is highlighted in the tree.

If the user then uses controls on the right side of the view to navigate elsewhere (e.g. click on a semantic link to go to a different model or class), the highlight on the tree does not change.

This is a bit disorientating. It would be good if the tree would always indicate the currently displayed item.

**Comment by pjmonks** [comment]
I think this is more complicated than it first appears. Whenever any link is clicked which would trigger a transition to another detail view, there is not enough information in that transition to know where in the tree that item is, or even if that item is loaded into the tree yet (due to not being fully expanded).

I think the implementation steps to add this change would be:

- [ ] Have a new endpoint added to `mdm-resources` - request the _ancestors_ of a given item with a known domain type and ID.
- [ ] Update the `UIRouter` config in `app-routing.module.ts`. Add a new state transition hook for `onSuccess()` to capture navigation changes.
- [ ] In the state transition hook, grab the catalogue item domain type and item ID from the URL/state parameters (if available), and broadcast an event with this data
- [ ] In the `mdm-folders-tree` component, subscribe to this broadcast event to capture the item navigated to.
- [ ] Use the new endpoint to request the ancestors of the item in current view
- [ ] Use the ancestor list to recursively expand the tree until the current item is found, then select it.

All of this would first be dependent on MauroDataMapper/mdm-core providing a tree item ancestor endpoint.

**Comment by jamesrwelch** [comment]
I think this is a key usability feature though, so I'd like to bump up its priority... completely understand we need the back-end endpoints first

**Comment by pjmonks** [comment]
I've noticed that MauroDataMapper/mdm-core#103 has been completed now, meaning this front-end task is now unblocked.

**Comment by pjmonks** [comment]
I have spent some time working on the implementation of this enhancement but have been having issues getting it to work. In theory, the process of automatically expanding the tree down to the required displayed catalogue item is sound, but I have encountered technical issues with the existing catalogue tree component that make this harder.

One primary issue is the loading of the tree and determining when to expand and fetch more child nodes. There is a mixture of `async/await` (or `Promises`) wrapped inside `Observables` for fetching data from the back-end (an observation noted in other areas of the code base). These existing mechanisms make it harder each time to adjust and add new features on top of, as following the process of fetching tree data has become very complex. 

The components have also organically grown in features and code complexity, making it harder to maintain. This will become even more problematic when future planned features, such as federation, come along which want to expand the tree view even further.

Based on this, my suggestion is to:

1. Put this particular issue on hold.
2. Perform a full review of the current feature set of the catalogue tree view, plus ideal new features to include (such as this one).
3. Plan a rewrite of the tree view components to better support long-term (using YouTrack).
4. Review the proposed task list and plan/prioritise the work.

For reference, there are a couple of issues already raised which could be considered for the rewrite:

* https://metadatacatalogue.myjetbrains.com/youtrack/issue/MC-9347
* https://metadatacatalogue.myjetbrains.com/youtrack/issue/MC-9615

**Comment by pjmonks** [comment]
The PR #649 has introduced a new tree view component that is more simplified, the key part is that the code to fetch more child nodes on expand is separated into a custom `DataSource` which makes the code much simpler to follow (and removes a lot of issues I've highlighted in previous comments). Once that PR is merged, it might be worthwhile revisiting this issue to see if it helps.
